diff --git a/BoardConfig.mk b/BoardConfig.mk
index d4b60c4..8d44d38 100644
--- a/BoardConfig.mk
+++ b/BoardConfig.mk
@@ -33,3 +33,7 @@ TARGET_OTA_ASSERT_DEVICE := galaxys2,i9100,GT-I9100,GT-I9100M,GT-I9100P,GT-I9100
 
 # Use the non-open-source parts, if they're present
 -include vendor/samsung/i9100/BoardConfigVendor.mk
+
+# BrickbugAftermath: Enable non-secure wiping of eMMC partitions in recovery
+BOARD_SUPPRESS_EMMC_WIPE := false
+COMMON_GLOBAL_CFLAGS += -DNO_SECURE_DISCARD
