diff --git a/BoardCommonConfig.mk b/BoardCommonConfig.mk
index 63042df..7e34f15 100644
--- a/BoardCommonConfig.mk
+++ b/BoardCommonConfig.mk
@@ -161,1 +161,5 @@ BOARD_UMS_LUNFILE := "/sys/class/android_usb/android0/f_mass_storage/lun0/file"
-BOARD_SUPPRESS_EMMC_WIPE := true
+# BrickbugAftermath:
+# Revert suppression of partition wiping in recovery. Use the safe, non-secure
+# discard command variant in case this code is ever run under an unsafe kernel.
+#BOARD_SUPPRESS_EMMC_WIPE := true
+COMMON_GLOBAL_CFLAGS += -DNO_SECURE_DISCARD
