diff --git a/BoardCommonConfig.mk b/BoardCommonConfig.mk
index 63042df..7e34f15 100644
--- a/BoardCommonConfig.mk
+++ b/BoardCommonConfig.mk
@@ -158,7 +158,11 @@ BOARD_UMS_LUNFILE := "/sys/class/android_usb/android0/f_mass_storage/lun0/file"
 BOARD_USES_MMCUTILS := true
 BOARD_HAS_NO_MISC_PARTITION := true
 BOARD_HAS_NO_SELECT_BUTTON := true
-BOARD_SUPPRESS_EMMC_WIPE := true
+# BrickbugAftermath:
+# Revert suppression of eMMC partition wiping in recovery.
+#BOARD_SUPPRESS_EMMC_WIPE := true
+# Explicit non-secure wiping is not required given Samsung's brickbug fix.
+#COMMON_GLOBAL_CFLAGS += -DNO_SECURE_DISCARD
 BOARD_RECOVERY_SWIPE := true
 TARGET_RECOVERY_FSTAB := device/samsung/galaxys2-common/rootdir/fstab.smdk4210
 RECOVERY_FSTAB_VERSION := 2
