diff --git a/drivers/mmc/host/mshci.c b/drivers/mmc/host/mshci.c
index f93a7eb..6ce9987 100644
--- a/drivers/mmc/host/mshci.c
+++ b/drivers/mmc/host/mshci.c
@@ -2046,7 +2046,17 @@ int mshci_add_host(struct mshci_host *host)
 	mmc->f_min = 400000;
 	mmc->f_max = host->max_clk;
 #ifdef CONFIG_MACH_U1
-	mmc->caps |= MMC_CAP_SDIO_IRQ;
+	/*
+	 * BrickbugAftermath:
+	 * Revert suppression of ERASE/TRIM/DISCARD eMMC commands.
+	 * 
+	 * Current kernel bugfix status:
+	 * -Brickbug: has fix
+	 * -MAG2GA TRIM bug: has fix
+	 * -Wear Leveling bug: HAS NO FIX
+	 */
+	/* mmc->caps |= MMC_CAP_SDIO_IRQ; */
+	mmc->caps |= MMC_CAP_SDIO_IRQ | MMC_CAP_ERASE;
 #else
 	mmc->caps |= MMC_CAP_SDIO_IRQ | MMC_CAP_ERASE;
 #endif
