diff --git a/drivers/mmc/host/mshci.c b/drivers/mmc/host/mshci.c
index f93a7eb..6ce9987 100644
--- a/drivers/mmc/host/mshci.c
+++ b/drivers/mmc/host/mshci.c
@@ -2046,7 +2046,17 @@ int mshci_add_host(struct mshci_host *host)
 	mmc->f_min = 400000;
 	mmc->f_max = host->max_clk;
 #ifdef CONFIG_MACH_U1
-	mmc->caps |= MMC_CAP_SDIO_IRQ;
+	/*
+	 * BrickbugAftermath:
+	 * Revert suppression of ERASE/TRIM/DISCARD eMMC commands.
+	 * 
+	 * Current kernel bugfix status:
+	 * -Brickbug: has fix
+	 * -MAG2GA TRIM bug: has fix
+	 * -Wear Leveling bug: HAS NO FIX
+	 */
+	/* mmc->caps |= MMC_CAP_SDIO_IRQ; */
+	mmc->caps |= MMC_CAP_SDIO_IRQ | MMC_CAP_ERASE;
 #else
 	mmc->caps |= MMC_CAP_SDIO_IRQ | MMC_CAP_ERASE;
 #endif
diff --git a/kernel/module.c b/kernel/module.c
index c99de29..2424b22 100644
--- a/kernel/module.c
+++ b/kernel/module.c
@@ -2431,9 +2431,17 @@ static int check_modinfo(struct module *mod, struct load_info *info)
 		if (err)
 			return err;
 	} else if (!same_magic(modmagic, vermagic, info->index.vers)) {
-		printk(KERN_ERR "%s: version magic '%s' should be '%s'\n",
+    	/*
+    	 * Ignore kernel module version magic mismatches.
+    	 * This hack is needed to make the patched kernel load
+    	 * kernel modules from the official CyanogenMod build.
+    	 */
+		/* printk(KERN_ERR "%s: version magic '%s' should be '%s'\n",
+		       mod->name, modmagic, vermagic);
+		return -ENOEXEC; */
+		printk(KERN_WARNING "%s: ignoring version magic mismatch,"
+		       " version magic '%s' should be '%s'\n",
 		       mod->name, modmagic, vermagic);
-		return -ENOEXEC;
 	}
 
 	if (get_modinfo(info, "staging")) {
